sms_alternative_provider/models/ir_sms_gateway.py

```diff
@@ -0,0 +1,183 @@
+
+        messages is a list of dictionaries [{
+            'id': sms.sms id,
+            'number': phone number,
+            'content': sms content,
+        }]
+
+        Returns list of dictionaries [{
+            'id': sms.sms id,
+            'state': sms.sms#state,
+            'failure_type': sms.sms#failure_type,
+            'sms_gateway_id': ir.sms.gateway id
+        }]
+        """
+        SmsSms = self.env["sms.sms"]
+        result = []
+        providers = self._send_get_providers(messages)
+        provider2messages = providers._send_partition_providers(messages)
+        for provider, messages_to_send in provider2messages.items():
+            if not provider and raise_exception:
+                raise exceptions.UserError(
+                    _("No suitable provider found for messages %s") % messages_to_send
+                )
+            provider_result = provider.send(messages_to_send) or []
+            sms = SmsSms.browse(
+                filter(None, map(operator.itemgetter("id"), provider_result))
+            )
+            sms.write({"sms_gateway_id": provider.id})
+            if handle_results:
+                provider._handle_results(provider_result)
+            result.extend(provider_result)
+        return result
+
+    @api.model
+    def _send_get_providers(self, messages):
+        """
+        Return all providers potentially suitable for the current context
+        """
+        return self.search(
+            ["|", ("company_id", "=", False), ("company_id", "=", self.env.company.id)]
+        )
+
+    def _send_partition_providers(self, messages):
+        """
+        Return a dict with providers in self as keys and lists of messages a provider
+        will handle as value
+        ie
+        {
+            ir.gateway.record(42,): [{message1}, {message2}],
+            ir.gateway.record(43,): [{message3}],
+            ir.gateway.record(44,): [],
+            ir.gateway.record(): [{messages not handled by any available provider}],
+        }
+        """
+        result = {}
+        remaining_messages = messages[:]
+        while remaining_messages:
+            message = remaining_messages.pop()
+            for this in self:
+                if this._can_send(message):
+                    result.setdefault(this, []).append(message)
+                    break
+            else:
+                result.setdefault(self.browse([]), []).append(message)
+        return result
+
+    def _can_send(self, message):
+        """
+        Determine if the provider can send a message
+        """
+        self.ensure_one()
+        if not self.prefix:
+            return True
+        return any(
+            (message.get("number") or "").startswith(prefix)
+            for prefix in self.prefix.split()
+        )
+
+    def _handle_results(self, results, unlink_failed=False, unlink_sent=True):
+        """
+        Write state of sms.sms objects based on results
```

Comment:
It might be nice to describe here which results are passed.

Also, are any of the original message key/value pairs passed? Would be handy to be able to rely on that.

```
            'id': sms.sms id,
            'number': phone number,
            'content': sms content,
```

Reply:
I now pass the whole list of messages as passed to _send, seems most general to me